research/minigo/minigo.py

```diff
@@ -0,0 +1,406 @@
+# Copyright 2018 MLBenchmark Group. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Train MiniGo with several iterations of RL learning.
+
+One iteration of RL learning consists of bootstrap, selfplay, gather and train:
+  bootstrap: Initialize a random model
+  selfplay: Play games with the latest model to produce data used for training
+  gather: Group games played with the same model into larger files of tfexamples
+  train: Train a new model with the selfplay results from the most recent
+    N generations.
+After training, validation can be preformed on the holdout data. Besides, given
+two models, evaluation can be applied to choose a stronger model.
+The training pipeline consists of multiple RL learning iterations to achieve
+better models.
+"""
+
+# pylint: disable=g-bad-import-order
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import os
+import random
+import sys
+import shutil
+import socket
+import time
+from tqdm import tqdm
+
+import tensorflow as tf
+
+import dual_net
+import evaluation
+import model_utils
+import preprocessing
+import selfplay_mcts
+import utils
+
+
+def _ensure_dir_exists(directory):
+  """Check if directory exists. If not, create it.
+
+  Args:
+    directory: A given directory
+  """
+  if directory.startswith('gs://'):  # for cloud logging
+    return
+  tf.gfile.MakeDirs(directory)
+
+
+def bootstrap(estimator_working_dir, models_dir, params):
+  """Initialize the model with random weights.
+
+  Args:
+    estimator_working_dir: tf.estimator working directory.
+    models_dir: Where to export the first bootstrapped generation.
+    params: Hyper params of the model.
+  """
+  bootstrap_name = model_utils.generate(0)
+  bootstrap_model_path = os.path.join(models_dir, bootstrap_name)
+
+  _ensure_dir_exists(estimator_working_dir)
+  _ensure_dir_exists(os.path.dirname(bootstrap_model_path))
+
+  print('Bootstrapping with working dir %s\n Model 0 exported to %s' %
+        (estimator_working_dir, bootstrap_model_path))
+  dual_net.bootstrap(estimator_working_dir, params)
+  dual_net.export_model(estimator_working_dir, bootstrap_model_path)
+
+
+def selfplay(model_name, models_dir, selfplay_dir, holdout_dir, sgf_dir,
```

Comment:
Can we pass in the network itself here, instead of the model_name? There is significant overhead (~33%) if we load the network for every game versus caching it.

Reply:
Do you mean we should load the network into memory in the main entry, and pass the network as a parameter to selfplay or other sub-functions that take the network as an input? This is the same like 'network caching', right?